숫자야구 [STEP 2] uemu, Charles

Charl-es opened the pull request:
@serena0720 

안녕하세요 세라나, STEP 2 제출합니다.
이번 스텝도 잘 부탁드립니다🙏
---

## 🤔고민했던 점
- 사용자에게 입력받고 그 수를 `userNumbers` 정수형 배열에 초기화하기 위해서 어떤 식으로 넣어줄 지를 고민했었습니다.
    * 사용자 입력을 받는 `readLine()` 메서드는 `String?` 타입이므로 먼저 `if let` 바인딩으로 옵셔널을 언래핑 했습니다.
    * 그 다음 입력받은 세 개의 수를 공백을 기준으로 나눠야겠다는 생각으로 `split(separator:)`메서드를 사용하여 공백을 기준으로 나눠 `[String]` 타입을 반환 받았습니다.
    *  `userNumbers`가 `[String]` 타입이므로 `String`을 `Int`로 변환을 하기위해 `compactMap` 고차함수를 사용하였고 클로저 내의 `Int()` 로 타입변환을 하여 `[Int]` 형 배열로 구현하였습니다.
        * 여기서 `compactMap` 고차함수를 사용한 이유가 `Int()` 로 타입변환을 시도하면 숫자가 아닌 문자같은 경우는 변환이 불가하기 때문에 옵셔널이 되기 때문입니다. `compactMap`는 기존의 배열의 각 요소를 새롭게 매핑을 하여 변형을 하나, 옵셔널 요소 또한 제거를 하고 새로운 배열을 리턴하는 고차함수 입니다.

- 사용자가 승리했을 경우와, 컴퓨터가 승리했을 경우에 게임이 다시 시작되지 않고 종료가 되는 현상을 고민했었습니다.
    - 이는 각 `if` 안에 `conductMenu()`를 호출함으로 해결하였습니다.

- 게임이 끝나고 새로운 게임을 시작했을 때, 컴퓨터의 임의의 수가 전 게임과 동일한 현상을 발견하여 이를 고민했습니다.
   - 이 또한 사용자와 컴퓨터가 승리했을 시 새로운 게임이 시작되기 전에 다시 새로운 값을 초기화 시켜주어 해결했습니다.  

- 게임이 끝나고 새로운 게임을 시작했을 때, 남은 기회가 초기화가 되지 않는 현상을 발견하여 이를 고민했습니다.
   - 전역변수였던 `remainsChance` 를 `playGame()` 함수 안에 `while` 문이 돌기전에 3으로 선언하여 해결하였습니다.

- 사용자 또는 컴퓨터가 승리하여 게임이 종료된 후, 사용자 메뉴에서 게임종료를 선택함에 있어 게임이 종료가 되지 않는 현상이 발견하여 이를 고민했습니다.
    - 이는 `if` 문 내에 `return`키워드를 사용하여 해결하였습니다.
---

## 🤩조언을 얻고 싶은 부분
- 구현함에 있어서 저희 팀은 함수 내에 꼭 필요한 함수를 호출한 것이지만, 함수 내에서는 최대한 함수를 호출하는 것을 지양해야 하는 지가 궁금합니다.

ue-mu committed:
Feat: 사용자 메뉴를 출력하고 입력받는 함수와 게임 숫자를 입력받는 함수 구현

Charl-es committed:
Feat: 유효성 검증 함수 구현

사용자가 입력하는 세 개의 숫자의 유효성을 검증하는 함수를 구현

ue-mu committed:
Feat: 시도 횟수 수정 및 유효성 함수 수정

시도 횟수3->9회로 변경

유효성 함수에서 중복숫자 로직 수정 및 입력할 수 있는 수의 범위 1부터 9까지 구현

serena0720 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -1,13 +1,12 @@
+var randomComputerNumbers = generateRandomNumbers()
 var userNumbers = [Int]()
 
 func generateRandomNumbers() -> [Int] {
```
30번째 줄

`If` 구문과 `guard` 구문의 조건 컨벤션 통일 고민이 필요한 거 같습니다.
저의 경우 가독성을 위해 조건 별로 개행을 하는 스타일을 갖고 있습니다.
하지만 아래에 첨부드리는 공식문서의 예시를 보면 조건문을 개행하지 않는 것을 볼 수 있습니다.
어떠한 컨벤션이던 본인의 스타일과 이유를 갖으면 된다고 생각합니다.
하지만 컨벤션 스타일을 의논하여 결정하신 후 통일성은 유지 부탁드립니다.
[Apple Developer - Optional Binding](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/thebasics/#Optional-Binding)
```swift
if let firstNumber = Int("4"), let secondNumber = Int("42"), firstNumber < secondNumber && secondNumber < 100 {
    print("\(firstNumber) < \(secondNumber) < 100")
}
```

Charl-es replied:
해당 부분은 `guard`문으로 통일하였습니다.

serena0720 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -49,26 +48,82 @@ func calculateStrikeCount(userNumbers: [Int]) -> Int {
+            return
+        }
+        
+        print("남은 기회 : \(remainsChance) ")
```
조건을 충족시키기 위해선 이 함수의 위치를 재고해봐야할 거 같습니다.😊

Charl-es replied:
반영하여 수정하였습니다!

serena0720 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -49,26 +48,82 @@ func calculateStrikeCount(userNumbers: [Int]) -> Int {
+        }
+        
+        print("남은 기회 : \(remainsChance) ")
+        inputGameNumbers()
```
107번째줄 코멘트와 함께 함수 호출 위치를 고민해보시면 좋을 거 같습니다.😊

serena0720 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -49,26 +48,82 @@ func calculateStrikeCount(userNumbers: [Int]) -> Int {
+    }
+}
+
+func conductMenu() {
```
`conductMenu`라는 네이밍이 다소 낯섭니다. `Menu`를 `시작`하는 함수라는 점을 고려하여 다시 네이밍해보시면 좋을 거 같습니다!👀

Charl-es replied:
해당 부분도 수정하였습니다!

serena0720 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -49,26 +48,82 @@ func calculateStrikeCount(userNumbers: [Int]) -> Int {
+}
+
+func conductMenu() {
+    print("1. 게임시작")
+    print("2. 게임종료")
+    print("원하는 기능을 선택해주세요 : ", terminator: "")
```
`print`문을 여러번 호출하는 것보다 하나의 `print`문으로 합치면 보다 코드가 간결해질 거 같습니다

Charl-es replied:
`멀티라인 스트링 리터럴` 을 이용하여 간결하게 수정하였습니다.

serena0720 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -49,26 +48,82 @@ func calculateStrikeCount(userNumbers: [Int]) -> Int {
+    }
+}
+
+func inputGameNumbers() {
```
`conductMenu`와 기능분리가 잘 된 거 같습니다! 좋은 함수 분리라고 생각합니다.

serena0720 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -49,26 +48,82 @@ func calculateStrikeCount(userNumbers: [Int]) -> Int {
+}
+
+func inputGameNumbers() {
+    while true {
```
`step1`의 리뷰와 겹치는 부분입니다.
`While`문의 조건 다시한번 생각해주시길 바랍니다😅

Charl-es replied:
조언해 주신대로 토글을 할 수 있는 `Bool` 타입의 변수를 선언하여 수정하였습니다.

serena0720 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -49,26 +48,82 @@ func calculateStrikeCount(userNumbers: [Int]) -> Int {
+    while true {
+        print("숫자 3개를 띄어쓰기로 구분하여 입력해주세요.")
+        print("중복 숫자는 허용하지 않습니다.")
+        print("입력 : ", terminator: "")
```
위와 동일하게 `print` 함수를 줄일 수 있을 거 같습니다.

serena0720 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -49,26 +48,82 @@ func calculateStrikeCount(userNumbers: [Int]) -> Int {
+        print("중복 숫자는 허용하지 않습니다.")
+        print("입력 : ", terminator: "")
+        
+        if let userInputNumbers = readLine() {
```
`inputGameNumbers`함수는 `conductMenu`의 `readline`을 받아서 사용하는 것이 충분히 가능해보입니다. 만약 `readline`을 받아서 사용하게 되면 불필요한 옵셔널 바인딩을 제거할 수 있을 거 같습니다.

Charl-es replied:
조언 주신 부분은 이해하였고, 차후 코드 구현에 신경쓰도록 하겠습니다.
~~가슴은 이해를 하지 못하였습니다~~

serena0720 replied:
미팅으로 말씀드린 부분들 복습 꼭 하시길 바랍니다!

serena0720 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -49,26 +48,82 @@ func calculateStrikeCount(userNumbers: [Int]) -> Int {
 
-        print("남은 기회 : \(remainsChance) ")
+func validateInput(userInput: String) {
+    userNumbers = userInput.split(separator: " ").compactMap({ Int($0) }).filter {$0 < 10}
```
`split`과 `compactMap`, `filter`를 적절하게 잘 사용하셨습니다!👍

추가로 `compactMap`과 `map`의 차이는 고려해보셨을까요?
두 메서드는 어떠한 차이를 갖을까요?

Charl-es replied:
`compactMap`는 기존의 배열의 각 요소를 새롭게 매핑을 하여 변형을 하나, 옵셔널 요소 또한 제거를 하고 새로운 배열을 리턴하는 고차함수 입니다.

`map` 또한 기존의 배열의 각 요소를 새롭게 매핑을 하여 새로운 배열을 리턴을 하는 고차함수이나, 클로저 내에서 옵셔널을 벗겨내는
로직을 구현해주지 않습니다.

고차함수 내부에서 새로운 로직을 구현해주지 않아도 옵셔널 요소를 제거해주는 `compactMap`을 사용하였습니다.

serena0720 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -49,26 +48,82 @@ func calculateStrikeCount(userNumbers: [Int]) -> Int {
+func validateInput(userInput: String) {
+    userNumbers = userInput.split(separator: " ").compactMap({ Int($0) }).filter {$0 < 10}
+    
+    guard userInput.isEmpty == false,
+         userInput.split(separator: " ").compactMap({ Int($0) }).count == 3,
+         Set(userNumbers).count > 2 else {
```
조건문 컨벤션 통일 고민해보시길 바랍니다!😊

serena0720 reviewed (changes requested):
안녕하세요 우에무, 찰스!
리뷰 남깁니다😊

### 고민했던 점
각 조건 별로 두분께서 많은 고민을 한 것이 느껴집니다! 제가 더이상 여쭤볼 부분이 없을 듯 합니다😊 이유있는 코드를 짜고자 고민하신 것이 느껴져서 앞으로의 행보가 더욱 기대되는 부분입니다!

### 조언을 얻고 싶은 부분
함수내에 다른 함수를 호출하는 것은 자주 통용되는 방법입니다. 다만 이전 리뷰에도 말씀드린 거처럼 함수의 역할과 기능에 맞게 호출을 해야한다는 점을 꼭 기억하시길 바랍니다. 단순히 기능의 흐름 만을 따라서 함수 호출을 하게 되면 한쪽에 기능이 과대해지거나, 불필요한 로직이 생기기 쉽습니다.
이 부분은 이번 리뷰를 수정해보시면서 경험해보실 수 있지 않을까싶습니다!
전체적인 흐름과 불필요한 로직을 찾는 것은 저도 아직 어려운 부분이기 때문에 끝나지 않는 숙제와도 같습니다. 현재 프로젝트에서 이미 완성도 있는 코드를 구현하신 여러분이기에 전체적인 흐름을 읽어가는 능력을 분명 빠르게 키우실 수 있을 거라 생각합니다.

--------

## ⭐️중요⭐️
항상 조건을 꼼꼼하게 체크하는 버릇을 들이시길 바랍니다! 그리고 PR전에는 꼭 함수 컨벤션 확인 및 프로젝트 실행 동작 확인은 필수적인 부분입니다!
<img src = "https://cdn.discordapp.com/attachments/1101039125936742460/1146801228165160990/2023-08-31_10.38.34.png" width= 350>
- 컴퓨터 승리 시 남은 기회 프린트 조건이 빠졌습니다. 정상적인 실행 동작 및 잘못된 실행 동작 예시 모두 항상 꼼꼼하게 확인해주세요!

----------

### 칭찬할 부분 👍

- 사용자가 승리 시 조건 잘 충족해 주셨습니다
<img src = "https://cdn.discordapp.com/attachments/1101039125936742460/1146801280950485024/2023-08-31_10.38.48.png" width= 350>

- 컴퓨터가 승리 시 조건 및 게임종료 조건 잘 충족해 주셨습니다.
<img src = "https://cdn.discordapp.com/attachments/1101039125936742460/1146801323828842637/2023-08-31_10.38.59.png" width= 350>

Charl-es committed:
Feat: conductMenu 네이밍 변경, isRunning 변수 추가, 컨벤션 통일, 코드 간결화

conductMenu -> startMenu 네이밍 변경.
while문 조건 토글을 위한 isRunning 변수 추가.
조건문 컨벤션 통일, print 함수 간결화

ue-mu committed:
Feat: 조건문 개행 수정

조건문 컨벤션 통일
